File: backend/recipe_objects/Ingredient.js
// Ingredient.js

// Define a set of valid units.
// You can adjust this array as needed.
const VALID_UNITS = ["lb", "oz", "cups", "tablespoon", "teaspoon"];

class Ingredient {
  /**
   * @param {string} name - The name of the ingredient.
   * @param {number} price - The price per unit of the ingredient.
   * @param {string} unit - The measuring unit (e.g., "lb", "cups").
   */
  constructor(name, price, unit) {
    this.name = name;
    this.price = price;
    this.unit = unit;
  }

  /**
   * Returns a string reporting the quantity with unit and name.
   * @param {number} quantity
   * @returns {string}
   */
  reportWithQuantity(quantity) {
    return `${quantity} ${this.unit} of ${this.name}`;
  }

  /**
   * Converts the Ingredient object into a plain object (for JSON serialization).
   * @returns {Object}
   */
  toDict() {
    return {
      name: this.name,
      price: this.price,
      unit: this.unit
    };
  }

  /**
   * Create an Ingredient object from a plain object.
   * @param {Object} data
   * @returns {Ingredient}
   */
  static fromDict(data) {
    return new Ingredient(data.name, data.price, data.unit);
  }

  /**
   * Calculates the total cost for the given quantity.
   * @param {number} quantity
   * @returns {number}
   */
  getCost(quantity) {
    return Math.round(this.price * quantity * 100) / 100;
  }

  /**
   * Returns a string representation of the Ingredient.
   * @returns {string}
   */
  represent() {
    return `Ingredient(name='${this.name}', price=${this.price})`;
  }

  getName() {
    return this.name;
  }

  setName(name) {
    if (typeof name === "string" && name.trim().length > 0) {
      this.name = name;
    } else {
      throw new Error("Invalid ingredient name.");
    }
  }

  getPrice() {
    return this.price;
  }

  setPrice(price) {
    if (typeof price === "number" && price >= 0) {
      this.price = price;
    } else {
      throw new Error("Price must be a non-negative number.");
    }
  }

  getUnit() {
    return this.unit;
  }

  setUnit(unit) {
    if (VALID_UNITS.includes(unit)) {
      this.unit = unit;
    } else {
      throw new Error(`Invalid unit. Choose from: ${VALID_UNITS.join(", ")}`);
    }
  }
}

// Export the VALID_UNITS as a property if needed elsewhere.
Ingredient.VALID_UNITS = VALID_UNITS;

module.exports = Ingredient;

File: backend/recipe_objects/Ingredient.py
from backend.Unit import Unit

class Ingredient:

    # name is the name of the ingredient
    # price is the price of the ingredient
    # unit is the measuring unit of the ingredient (lb, oz, cups, tablespoon, etc.)
    def __init__(self, name, price, unit):
        self.name = name
        self.price = price
        self.unit = unit

    def report_with_quantity(self, quantity):
        return f"{quantity} {self.unit} of {self.name}"

    def to_dict(self):
        """Convert Ingredient object to a dictionary for JSON serialization."""
        return {
            "name": self.name,
            "price": self.price,
            "unit": self.unit
        }

    @classmethod
    def from_dict(cls, data):
        """Create an Ingredient object from a dictionary."""
        return cls(data["name"], data["price"], data["unit"])

    def get_cost(self, quantity):
        return round(self.price * quantity, 2)
    
    def represent(self):
        return f"Ingredient(name='{self.name}', price={self.price})"
    
    def get_name(self):
        return self.name
    
    def set_name(self, name):
        if isinstance(name, str) and name.strip():
            self.name = name
        else:
            raise ValueError("Invalid ingredient name.")
    
    def get_price(self):
        return self.price
    
    def set_price(self, price):
        if isinstance(price, (int, float)) and price >= 0:
            self.price = price
        else:
            raise ValueError("Price must be a non-negative number.")
        
    def get_unit(self):
        return self.unit
    
    def set_unit(self, unit):
        if unit in Unit.VALID_UNITS:
            self.unit = unit
        else:
            raise ValueError(f"Invalid unit. Choose from: {', '.join(Unit.VALID_UNITS)}")
    

    
